packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -1,43 +1,192 @@
 import { makeTracer } from '@agoric/internal';
-import type { OrchestrationAccount } from '@agoric/orchestration';
 import type { Zone } from '@agoric/zone';
 import { Fail, q } from '@endo/errors';
+import { M } from '@endo/patterns';
+import { VowShape } from '@agoric/vow';
+import { atob, decodeBase64 } from '@endo/base64';
+import { decodeAbiParameters } from 'viem';
+import { type MapStore } from '@agoric/store';
+import type {
+  AxelarGmpIncomingMemo,
+  SupportedDestinationChains,
+} from '@agoric/orchestration/src/axelar-types.js';
+import type { VTransferIBCEvent } from '@agoric/vats';
+import type { FungibleTokenPacketData } from '@agoric/cosmic-proto/ibc/applications/transfer/v2/packet.js';
+import type { OrchestrationAccount } from '@agoric/orchestration';
 import { YieldProtocol } from './constants.js';
 
-const interfaceTODO = undefined;
-
 const trace = makeTracer('PortExo');
 
-type Accounts = {
-  Aave: never; // TODO
-  Compound: never; // TODO
-  USDN: OrchestrationAccount<{ chainId: 'noble-any' }>;
+const KeeperI = M.interface('keeper', {
+  init: M.call(M.string(), M.remotable('OrchestrationAccount')).returns(),
+  getAccount: M.call(M.string()).returns(M.remotable('OrchestrationAccount')),
+});
+
+const EvmTapI = M.interface('EvmTap', {
+  receiveUpcall: M.call(M.record()).returns(M.or(VowShape, M.undefined())),
+});
+
+const supportedDestinationChains: SupportedDestinationChains[] = [
+  'Avalanche',
+  'Base',
+  'Ethereum',
+];
+
+type EVMProtocolState = {
+  localAccount?: OrchestrationAccount<{ chainId: 'agoric-any' }>;
+  remoteAccountAddress?: string;
+  evmChain?: string;
+  isActive: boolean;
+};
+
+type CosmosProtocolState = {
+  icaAccount?: OrchestrationAccount<{ chainId: 'noble-any' }>;
+  isActive: boolean;
+};
+
+type PortfolioKitStateShape = {
+  protocolStates: MapStore<
+    YieldProtocol,
+    CosmosProtocolState | EVMProtocolState
+  >;
 };
 
 export const preparePortfolioKit = (zone: Zone) =>
   zone.exoClassKit(
     'Portfolio',
-    interfaceTODO,
-    () =>
-      ({
-        USDN: undefined,
-      }) as Partial<Accounts>,
     {
+      keeper: KeeperI,
+      invitationMakers: M.interface('invitationMakers', {
+        // supplyToAave: M.call(M.record()).returns(M.promise()),
+        // borrowFromAave: M.call(M.record()).returns(M.promise()),
+        // supplyToCompound: M.call(M.record()).returns(M.promise()),
+        // borrowFromCompound: M.call(M.record()).returns(M.promise()),
+        // withdraw: M.call(M.record()).returns(M.promise()),
+      }),
+      tap: EvmTapI,
+    },
+
+    (): PortfolioKitStateShape => {
+      return harden({
+        protocolStates: zone.mapStore('protocolStates'),
+      });
+    },
+    {
+      tap: {
+        receiveUpcall(event: VTransferIBCEvent) {
+          trace('receiveUpcall', event);
```

Comment:
I'm starting to feel like adding a `trace` indicates that a test is in order. Sometimes it's OK to keep the trace even when we have the test. But I'd like us to think pretty hard about why we call trace. In this case, it's a signal that a message is coming in from outside the vat. That seems like something that will be useful in production monitoring, so this is a keeper.